File: go/tokenizer/tokenizer_test.go
package tokenizer_test

import (
	"os"
	"testing"

	"github.com/ailakshya/infergo/tokenizer"
)

const testTokenizerPath = "/tmp/bert_tokenizer/tokenizer.json"

func skipIfMissing(t *testing.T) {
	t.Helper()
	if _, err := os.Stat(testTokenizerPath); err != nil {
		t.Skipf("tokenizer not found at %s — run: python3 -c \"from transformers import AutoTokenizer; AutoTokenizer.from_pretrained('bert-base-uncased').save_pretrained('/tmp/bert_tokenizer')\"", testTokenizerPath)
	}
}

// ─── Load / Close ─────────────────────────────────────────────────────────────

func TestLoad(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	tok.Close()
}

func TestLoadBadPath(t *testing.T) {
	_, err := tokenizer.Load("/no/such/tokenizer.json")
	if err == nil {
		t.Fatal("expected error for bad path, got nil")
	}
}

func TestDoubleClose(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	tok.Close()
	tok.Close() // must not panic
}

// ─── VocabSize ────────────────────────────────────────────────────────────────

func TestVocabSize(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	if tok.VocabSize() <= 0 {
		t.Errorf("VocabSize() = %d, want > 0", tok.VocabSize())
	}
}

// ─── Encode ───────────────────────────────────────────────────────────────────

func TestEncode(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	enc, err := tok.Encode("Hello world", false, 512)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	if len(enc.IDs) == 0 {
		t.Error("Encode returned empty IDs")
	}
	if len(enc.IDs) != len(enc.AttentionMask) {
		t.Errorf("IDs len %d != AttentionMask len %d", len(enc.IDs), len(enc.AttentionMask))
	}
	for i, m := range enc.AttentionMask {
		if m != 1 {
			t.Errorf("attention_mask[%d] = %d, want 1", i, m)
		}
	}
}

func TestEncodeWithSpecialTokens(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	without, err := tok.Encode("Hello", false, 512)
	if err != nil {
		t.Fatalf("Encode without special: %v", err)
	}
	with, err := tok.Encode("Hello", true, 512)
	if err != nil {
		t.Fatalf("Encode with special: %v", err)
	}
	if len(with.IDs) < len(without.IDs) {
		t.Errorf("with special tokens (%d) shorter than without (%d)", len(with.IDs), len(without.IDs))
	}
}

func TestEncodeTruncates(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	enc, err := tok.Encode("the quick brown fox jumps over the lazy dog", false, 3)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	if len(enc.IDs) != 3 {
		t.Errorf("want 3 tokens, got %d", len(enc.IDs))
	}
}

func TestEncodeEmpty(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	_, err = tok.Encode("", false, 512)
	if err != nil {
		t.Errorf("Encode empty string: %v", err)
	}
}

// ─── Decode ───────────────────────────────────────────────────────────────────

func TestDecodeRoundTrip(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	enc, err := tok.Encode("hello world", false, 512)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	text, err := tok.Decode(enc.IDs, true)
	if err != nil {
		t.Fatalf("Decode: %v", err)
	}
	if text != "hello world" {
		t.Errorf("round-trip: got %q, want %q", text, "hello world")
	}
}

func TestDecodeEmpty(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	text, err := tok.Decode(nil, true)
	if err != nil {
		t.Fatalf("Decode nil: %v", err)
	}
	if text != "" {
		t.Errorf("Decode nil: got %q, want empty string", text)
	}
}

func TestDecodeMultipleSentences(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	sentences := []string{
		"hello world",
		"the quick brown fox",
		"inference at the speed of go",
	}
	for _, s := range sentences {
		enc, err := tok.Encode(s, false, 512)
		if err != nil {
			t.Fatalf("Encode %q: %v", s, err)
		}
		got, err := tok.Decode(enc.IDs, true)
		if err != nil {
			t.Fatalf("Decode %q: %v", s, err)
		}
		if got != s {
			t.Errorf("round-trip %q: got %q", s, got)
		}
	}
}

// ─── DecodeToken ──────────────────────────────────────────────────────────────

func TestDecodeToken(t *testing.T) {
	skipIfMissing(t)
	tok, err := tokenizer.Load(testTokenizerPath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	defer tok.Close()

	enc, err := tok.Encode("hello", false, 512)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	for _, id := range enc.IDs {
		piece, err := tok.DecodeToken(id)
		if err != nil {
			t.Fatalf("DecodeToken(%d): %v", id, err)
		}
		if piece == "" {
			t.Errorf("DecodeToken(%d) returned empty string", id)
		}
	}
}

File: cpp/tokenizer/tokenizer_test.cpp
#include "tokenizer.hpp"

#include <gtest/gtest.h>

#ifndef TEST_TOKENIZER_PATH
#define TEST_TOKENIZER_PATH "/tmp/bert_tokenizer/tokenizer.json"
#endif

using namespace infergo;

// ─── Construction ─────────────────────────────────────────────────────────────

TEST(TokenizerWrapper, ConstructValid) {
    EXPECT_NO_THROW(TokenizerWrapper t(TEST_TOKENIZER_PATH));
}

TEST(TokenizerWrapper, ConstructBadPathThrows) {
    EXPECT_THROW(TokenizerWrapper t("/no/such/tokenizer.json"), std::runtime_error);
}

TEST(TokenizerWrapper, MoveConstruct) {
    TokenizerWrapper a(TEST_TOKENIZER_PATH);
    EXPECT_NO_THROW(TokenizerWrapper b(std::move(a)));
}

TEST(TokenizerWrapper, MoveAssign) {
    TokenizerWrapper a(TEST_TOKENIZER_PATH);
    TokenizerWrapper b(TEST_TOKENIZER_PATH);
    EXPECT_NO_THROW(b = std::move(a));
}

// ─── vocab_size ───────────────────────────────────────────────────────────────

TEST(TokenizerWrapper, VocabSizePositive) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    EXPECT_GT(t.vocab_size(), 0);
}

// ─── encode ───────────────────────────────────────────────────────────────────

TEST(TokenizerWrapper, EncodeHelloWorld) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    auto enc = t.encode("Hello world", /*add_special_tokens=*/false);
    ASSERT_GT(static_cast<int>(enc.ids.size()), 0);
    EXPECT_EQ(enc.ids.size(), enc.attention_mask.size());
    for (int m : enc.attention_mask) EXPECT_EQ(m, 1);
}

TEST(TokenizerWrapper, EncodeWithSpecialTokens) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    auto without = t.encode("Hello", false);
    auto with    = t.encode("Hello", true);
    // With special tokens should be longer
    EXPECT_GE(with.ids.size(), without.ids.size());
}

TEST(TokenizerWrapper, EncodeTruncatesAtMaxTokens) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    auto full  = t.encode("the quick brown fox jumps over the lazy dog", false, 512);
    auto trunc = t.encode("the quick brown fox jumps over the lazy dog", false, 3);
    EXPECT_EQ(static_cast<int>(trunc.ids.size()), 3);
    for (int i = 0; i < 3; ++i) EXPECT_EQ(trunc.ids[i], full.ids[i]);
}

TEST(TokenizerWrapper, EncodeEmptyString) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    EXPECT_NO_THROW(auto enc = t.encode("", false));
}

// ─── decode ───────────────────────────────────────────────────────────────────

TEST(TokenizerWrapper, DecodeRoundTrip) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    // The key acceptance criterion for T-19
    auto enc = t.encode("hello world", false);
    ASSERT_GT(static_cast<int>(enc.ids.size()), 0);
    std::string decoded = t.decode(enc.ids, /*skip_special_tokens=*/true);
    EXPECT_EQ(decoded, "hello world");
}

TEST(TokenizerWrapper, DecodeEmptyIds) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    std::string result;
    EXPECT_NO_THROW(result = t.decode({}, true));
    EXPECT_TRUE(result.empty());
}

TEST(TokenizerWrapper, DecodeMultipleSentences) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    const std::vector<std::string> sentences = {
        "hello world",
        "the quick brown fox",
        "inference at the speed of go",
    };
    for (const auto& s : sentences) {
        auto enc = t.encode(s, false);
        ASSERT_GT(static_cast<int>(enc.ids.size()), 0) << "sentence: " << s;
        std::string decoded = t.decode(enc.ids, true);
        EXPECT_EQ(decoded, s) << "round-trip failed for: " << s;
    }
}

// ─── decode_token ─────────────────────────────────────────────────────────────

TEST(TokenizerWrapper, DecodeTokenReturnsNonEmpty) {
    TokenizerWrapper t(TEST_TOKENIZER_PATH);
    // Encode "hello" and decode each token individually
    auto enc = t.encode("hello", false);
    ASSERT_GT(static_cast<int>(enc.ids.size()), 0);
    for (int32_t id : enc.ids) {
        std::string piece;
        EXPECT_NO_THROW(piece = t.decode_token(id));
        // Each piece is a non-empty string for normal tokens
        EXPECT_FALSE(piece.empty()) << "empty piece for id=" << id;
    }
}
